### PR TITLE
Move to a callback-style approach to deserializing objects

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Followup to https://github.com/dotnet/roslyn/pull/72929

Instead of allocating a large, temporary, array just to pass values around, only to then add those values to a cache, now we just pass a callback to the deserializer to allow it to directly add to the cache.

Note: this also has the benefit of allowing the cache to be filled *as* we deserialize an item, versus having to deserialize them all.  This can help concurrent requests that then may find those items in teh cache, and not have to requery them.

--

Note: it's been a long standing goal of mine to make it so that we can read/write from those pipes without needing the intermediary stream.  that would then allow us to populate the cache on the OOP side *immediately* as objects come across, without having to wait for them all to be written.  I intend to make that possible this week.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
@ToddGrun ptal.  note: there is a followup PR to thsi that takes this idea and spreads it much further, avoiding a *lot* of intermediary list allocs.  THat PR also moves to a `TArg`-style model so we can do this work without allocating lambdas either.  But i wanted to start here to get the ball rolling on the main idea.

**Comment by CyrusNajmabadi** [comment]
https://github.com/dotnet/roslyn/pull/72944 is part 2 of this.  But shoudl not be reviewed until this goes in.

File: src/Workspaces/Remote/ServiceHub/Host/AssetProvider.cs [hunk 1]
```diff
@@ -238,18 +238,21 @@ static void AddAll(HashSet<Checksum> checksums, ChecksumCollection checksumColle
...
+                var currentIndex = 0;
+                await RequestAssetsAsync(assetPath, missingChecksumsMemory, (int index, T missingAsset) =>
                 {
-                    var missingChecksum = missingChecksums[i];
-                    var missingAsset = missingAssets[i];
+                    Contract.ThrowIfTrue(currentIndex != index);
+
+                    var missingChecksum = missingChecksums[index];
 
                     AddResult(missingChecksum, missingAsset);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/ServiceHub/Host/AssetProvider.cs:249]: there are several captures in this lambda.  but they will go away in followup pr.

File: src/Workspaces/Remote/Core/RemoteHostAssetSerialization.cs [hunk 2]
```diff
@@ -77,20 +77,19 @@ static void WriteAsset(ObjectWriter writer, ISerializerService serializer, Solut
...
-                return ReadData<T>(stream, solutionChecksum, objectCount, serializerService, cancellationToken);
+                ReadData<T>(stream, solutionChecksum, objectCount, serializerService, callback, cancellationToken);
             }
         }
 
-        public static ImmutableArray<T> ReadData<T>(Stream stream, Checksum solutionChecksum, int objectCount, ISerializerService serializerService, CancellationToken cancellationToken)
+        public static void ReadData<T>(
+            Stream stream, Checksum solutionChecksum, int objectCount, ISerializerService serializerService, Action<int, T> callback, CancellationToken cancellationToken)
         {
-            var results = new T[objectCount];
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/Core/RemoteHostAssetSerialization.cs:92]: this is the major alloc removed.

**Comment by ToddGrun** [comment]
Nice!

**Review by ToddGrun — APPROVED** [review]
:shipit:

**Comment by CyrusNajmabadi** [comment]
@jasonmalinowski 	For review when you get back.

**Pull request merged**